File: src/task4_artifacts.R
# Konfiguration und numerisch wirksame Citrus-Cluster ohne Zusatzpakete prüfen.

check_run_config <- function(config_path, expected, artifact_paths, run_training) {
  present <- file.exists(artifact_paths)
  if (!file.exists(config_path) && !any(present)) {
    if (!run_training) stop("Keine Ergebnisse vorhanden; zuerst einen Lauf berechnen.")
    return(FALSE)
  }
  if (!file.exists(config_path)) {
    stop("Konfigurationsnachweis fehlt: ", basename(config_path),
         ". Auch der Comparison benötigt einen passenden Nachweis. ",
         "Für einen neuen Lauf die alten Artefakte separat sichern und ",
         "aus den aktiven Ergebnispfaden verschieben; sie werden nicht überschrieben.")
  }
  if (!identical(readRDS(config_path), expected)) {
    stop("Gespeicherte Ergebnisse passen nicht zur aktuellen Konfiguration. ",
         "Alte Artefakte separat sichern und einen getrennten Lauf beginnen.")
  }
  if (!all(present)) {
    stop("Unvollständiger Zwischenstand: ",
         paste(basename(artifact_paths[!present]), collapse = ", "))
  }
  TRUE
}

effective_cluster_ids <- function(coefficients, tolerance) {
  stopifnot(is.finite(tolerance), tolerance >= 0, all(is.finite(coefficients)))
  feature_names <- rownames(coefficients)
  selected <- grepl("^cluster [0-9]+ abundance$", feature_names) &
    abs(coefficients[, 1]) > tolerance
  as.integer(sub("^cluster ([0-9]+) abundance$", "\\1", feature_names[selected]))
}

normalize_cluster_tables <- function(predictions, selection, clusters, tolerance) {
  # Auch alte, bereits gespeicherte Profile können ohne Neutraining bereinigt werden.
  stopifnot(all(is.finite(clusters$coefficient)))
  clusters <- clusters[abs(clusters$coefficient) > tolerance, , drop = FALSE]
  counts <- table(unique(clusters[c("split_id", "cluster_id")])$split_id)
  for (name in c("predictions", "selection")) {
    value <- get(name)
    count <- as.integer(counts[as.character(value$split_id)])
    count[is.na(count)] <- 0L
    value$selected_cluster_count <- count
    assign(name, value)
  }
  list(predictions = predictions, selection = selection, clusters = clusters)
}

validate_comparison_config <- function(root) {
  # Nur Nachweise lesen; keine FCS-Daten einlesen und keine Modelle fitten.
  tables <- file.path(root, "results", "tables")
  data <- file.path(root, "NK_cell_dataset", "NK_cell_dataset")
  path <- file.path(tables, "task4_citrus_predictions_gated_alive_full.config.rds")
  expected <- readRDS(path)
  expected$schema_version <- 1L
  parameters <- list(
    method = "citrus", gate = "gated_alive", run_mode = "full",
    transform_cofactor = 5, file_sample_size = 10000L,
    minimum_cluster_size_fraction = 0.0005, coefficient_tolerance = 1e-10,
    inner_folds = 3L, clustering = "hierarchical", feature_type = "abundances",
    model = "glmnet", alpha = 1, standardize_features = TRUE,
    selection = "cv.min", score_rounding_digits = 15L, decision_threshold = 0.5,
    citrus_commit = "d02baae544abdc403704aaceb75d1e7931a0331c",
    rclusterpp_commit = "a07380683ce7a6849af8ec27db6439ea3a707890"
  )
  expected$parameters[names(parameters)] <- parameters
  inputs <- c(file.path(tables, "task4_donor_splits.csv"),
              file.path(data, "NK_fcs_samples_with_labels.csv"),
              file.path(data, "NK_markers.csv"),
              sort(list.files(file.path(data, "NK_cell_dataset", "gated_alive"),
                              pattern = "\\.fcs$", full.names = TRUE)))
  expected$inputs <- setNames(as.character(tools::md5sum(inputs)), basename(inputs))
  notebook <- jsonlite::read_json(file.path(root, "notebooks", "04d_citrus.ipynb"))
  definitions <- new.env(parent = environment())
  eval(parse(text = paste0(unlist(notebook$cells[[9]]$source), collapse = "")), definitions)
  functions <- c(mget(c("read_citrus_files", "build_fixed_fold_clustering",
                       "extract_selected_cluster_profiles", "run_citrus_split"), definitions),
                 list(effective_cluster_ids, normalize_cluster_tables))
  expected$implementation <- unname(lapply(functions, function(f) paste(deparse(body(f)), collapse = "\n")))
  artifacts <- file.path(tables, paste0("task4_citrus_", c("predictions", "selection", "clusters"),
                                     "_gated_alive_full.csv"))
  check_run_config(path, expected, artifacts, run_training = FALSE)
  cat("Citrus-Konfiguration, Eingabeprüfsummen und Trainingscode stimmen.\n")
}

File: src/task4_artifacts.py
"""Konfiguration und Vollständigkeit lokaler Aufgabe-4-Zwischenstände prüfen."""

import hashlib
import json
from pathlib import Path

import numpy as np
import pandas as pd


def file_digest(path):
    """Datei ausschließlich lesen und ihren Inhalt mit SHA-256 kennzeichnen."""
    with Path(path).open("rb") as stream:
        return hashlib.file_digest(stream, "sha256").hexdigest()


def make_run_config(parameters, input_paths, notebook_path, code_cells):
    """Parameter, Eingangsdaten und relevante Notebook-Zellen festhalten.

    Split-Limits und erzwungene Wiederholungen ändern das Modellverfahren nicht
    und gehören nicht in parameters. Die gesamte Splitdatei wird dagegen gehasht.
    """
    notebook = json.loads(Path(notebook_path).read_text(encoding="utf-8"))
    source = "\n\n".join(
        "".join(notebook["cells"][index]["source"]) for index in code_cells
    )
    return {
        "schema_version": 1,
        "parameters": parameters,
        "inputs": {Path(path).name: file_digest(path) for path in input_paths},
        "implementation_sha256": hashlib.sha256(source.encode()).hexdigest(),
    }


def check_run_config(config_path, expected, artifact_paths, run_training):
    """Bei unvollständigen, alten oder anders konfigurierten Artefakten stoppen.

    Schreibt nichts. Ein neuer Konfigurationsnachweis wird erst zusammen mit
    dem ersten erfolgreich berechneten Split gespeichert.
    """
    config_path = Path(config_path)
    paths = [Path(path) for path in artifact_paths]
    present = [path.exists() for path in paths]
    if not config_path.exists() and not any(present):
        if not run_training:
            raise FileNotFoundError("Keine Ergebnisse vorhanden; zuerst einen Lauf berechnen.")
        return False
    if not config_path.exists():
        raise ValueError(
            f"Konfigurationsnachweis fehlt: {config_path.name}. "
            "Altbestände können in 04e historisch ausgewertet werden. "
            "Für einen neuen Methodenlauf die bisherigen Artefakte separat sichern "
            "und aus den aktiven Ergebnispfaden verschieben; sie werden nicht überschrieben."
        )
    stored = json.loads(config_path.read_text(encoding="utf-8"))
    if stored != expected:
        changed = sorted(key for key in set(stored) | set(expected)
                         if stored.get(key) != expected.get(key))
        raise ValueError(
            "Gespeicherte Ergebnisse passen nicht zur aktuellen Konfiguration "
            f"({', '.join(changed)}). Alte Artefakte separat sichern und "
            "für die neue Konfiguration einen getrennten Lauf beginnen."
        )
    if not all(present):
        missing = ", ".join(path.name for path in paths if not path.exists())
        raise ValueError(f"Unvollständiger Zwischenstand; fehlende Artefakte: {missing}")
    return True


def write_run_config(path, config):
    """Den tatsächlich verwendeten Konfigurationsnachweis speichern."""
    Path(path).write_text(json.dumps(config, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def validate_prediction_splits(predictions, donor_splits):
    """Nur vollständig gespeicherte äußere Spendersplits zum Fortsetzen zulassen."""
    columns = ["split_id", "split_seed", "donor_id", "y_true"]
    if predictions.empty or predictions.duplicated(["split_id", "donor_id"]).any():
        raise ValueError("Leerer oder doppelt gespeicherter Vorhersage-Zwischenstand.")
    expected = donor_splits.loc[
        donor_splits["outer_partition"].eq("test")
        & donor_splits["split_id"].isin(predictions["split_id"]),
        ["split_id", "split_seed", "donor_id", "label"],
    ].rename(columns={"label": "y_true"})
    matched = predictions[columns].merge(expected, on=columns, validate="one_to_one")
    if len(matched) != len(predictions) or len(matched) != len(expected):
        raise ValueError("Gespeicherte Vorhersagen sind unvollständig oder passen nicht zu den Splits.")
    if not np.isfinite(predictions[["score", "decision_threshold"]].to_numpy()).all():
        raise ValueError("Nicht-endliche Scores oder Entscheidungsschwellen.")
    if not predictions["y_pred"].eq(
        predictions["score"].ge(predictions["decision_threshold"]).astype(int)
    ).all():
        raise ValueError("Gespeicherte Klassen stimmen nicht mit Score und Schwelle überein.")


def validate_parameter_table(parameters, predictions, markers, group_columns, required_columns=()):
    """Pro Modell/Filter genau einen endlichen Parametersatz je Marker verlangen."""
    missing = set(required_columns) - set(parameters.columns)
    if missing:
        raise ValueError(f"Fehlende Modellparameterspalten: {', '.join(sorted(missing))}")
    if required_columns and not np.isfinite(parameters[list(required_columns)].to_numpy(dtype=float)).all():
        raise ValueError("Nicht-endliche gespeicherte Modellparameter.")
    if set(parameters["split_id"]) != set(predictions["split_id"]):
        raise ValueError("Modellparameter und Vorhersagen enthalten unterschiedliche Splits.")
    if parameters.duplicated([*group_columns, "marker"]).any():
        raise ValueError("Doppelte Marker in gespeicherten Modellparametern.")
    for _, group in parameters.groupby(group_columns):
        if len(group) != len(markers) or set(group["marker"]) != set(markers):
            raise ValueError("Unvollständige Marker in gespeicherten Modellparametern.")
    if not np.isfinite(parameters.select_dtypes(include="number").to_numpy()).all():
        raise ValueError("Nicht-endliche gespeicherte Modellparameter.")
